eager qat: refactor conv-bn to make conv with scaled weights a func

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* **#39837 eager qat: refactor conv-bn to make conv with scaled weights a func**
* #39840 qat graph mode: make `_forward` of per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script

Summary:

WIP, not ready for review yet.

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21998753](https://our.internmc.facebook.com/intern/diff/D21998753)

vkuzo committed:
eager qat: refactor conv-bn to make conv with scaled weights a func

Summary:

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -79,15 +79,27 @@ def freeze_bn_stats(self):
+    #   In particular, the inputs and outputs of this function should be
+    #   observed, and no intermediate values of this function should be
+    #   observed.
+    def _conv_forward_and_unscale(self, input, scaled_weight, scale_factor):
```
I actually mean an independent function. this is a method for the class

jerryzh168 replied:
rule for graph mode right now is we will go through CallMethod, but we will not go through the CallFunction in the list

vkuzo replied:
yep, we're on the same page.  Not intending to directly reuse this code, more structuring the eager mode code the same way we aligned for the graph mode IR, for easier testing during development.

let me mark this PR as wip, not ready for review yet.

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

WIP, not ready for review yet.

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21998753](https://our.internmc.facebook.com/intern/diff/D21998753)

[ghstack-poisoned]

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

WIP, not ready for review yet.

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21998753](https://our.internmc.facebook.com/intern/diff/D21998753)

[ghstack-poisoned]

vkuzo committed:
Update on "eager qat: refactor conv-bn to make conv with scaled weights a func"


Summary:

WIP, not ready for review yet.

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21998753](https://our.internmc.facebook.com/intern/diff/D21998753)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
+    const Tensor& input, const Tensor& scaled_weight,
+    const Tensor& scale_factor, const Tensor& bias, IntArrayRef stride,
+    IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  Tensor result = at::convolution(
```
why are we not using conv2d?

vkuzo replied:
I'm flexible, they eventually call the same thing.  Happy to change it.

jerryzh168 reviewed (approved):
LG, had one question

jerryzh168 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
+  Tensor result = at::convolution(
+      input, scaled_weight, bias, stride, padding, dilation, false, {{0, 0}},
+      groups);
+  result = result / scale_factor.reshape({1, -1, 1, 1});
```
also should we make this work for conv3d as well?

jerryzh168 replied:
btw fusion is already extended to work with conv3d: https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/fusion.py#L17

vkuzo replied:
yep, for sure.  Planning to do that in future PR, to keep PRs smaller.  Will add a TODO.

raghuramank100 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
+    const Tensor& scale_factor, const Tensor& bias, IntArrayRef stride,
+    IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  Tensor result = at::convolution(
+      input, scaled_weight, bias, stride, padding, dilation, false, {{0, 0}},
```
Are we adding the bias twice here? at::convolution takes in bias as an argument and later bias is added

raghuramank100 replied:
We do not need this op, we can call at::convolution directly, please see below

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -83,11 +83,21 @@ def _forward(self, input):
-        conv_orig = conv / scale_factor.reshape([1, -1, 1, 1])
-        if self.bias is not None:
-            conv_orig = conv_orig + self.bias.reshape([1, -1, 1, 1])
+
```
A comment on the original code: I dont follow the logic here: We seem to be calling conv_forward , which includes the bias, though the comment says that "this does not include the conv bias"

raghuramank100 replied:
Our original implementation is complicated and can be simplified, we first add a bias as part of the call to _conv_forward, Then we again add the bias to conv_orig. All of this is removed by the batch norm op as part of subtracting the batch mean.

vkuzo replied:
> # this does not include the conv bias

yeah, this comment was not describing the code well

raghuramank100 replied:
A cleaner implementation is to do:
``` python
scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))/scale_factor.reshape([-1,1,1,1])
F.convolution(input, scaled_weight, self.bias, self.stride, padding, self.dilation, self.groups)
```
```

vkuzo replied:
good point.  The code can potentially be made simpler with that refactor, thanks!

from what we discussed in chat, we'd still need a separate function for the above, because we want to insert the fake_quant on conv.weight directly at the op boundary of the aten op (to fit with how the rest of the code works and avoid custom logic).  So, since fake_quant is divided by scale_factor above, it wouldn't work out of the box.
